### ISSUE TITLE
Pandemic - player turn gets stuck in endless loop

### ISSUE THREAD

**Issue by kristian-angelin** [GAIGResearch/TabletopGames#157]
Player turn sometimes gets stuck in loop after infecting cities step. After infecting cities the turn goes back to perform player actions and lets one more action be executed before going to infecting cities again. The loop continues until the game is lost.

- The nextNode/childNode of InfectCities node (somehow) gets set to **null**, causing the AbstractRuleBasedForwardModel to reset the current node to the root node.

The issue appears when playing with different setups of MCTS players. Haven't tested with other types of players.
I'm looking into this problem but if anyone got any suggestions I would appriciate it.

**Comment by rdgain** [comment]
Thanks for bringing this up! 

I've had a look at this and it does seem there are several errors going on - the main one I've found is in the fact that when players go over the limit of cards, if they have to discard more than 1, several reactions for the player are added in the queue. However, the rule tree only handles one of these - leaving several reactions for the player in the queue to be picked up next time actions are processed and messing up turn order from that point on. The Resilient Population event card also seems to mess up turn order when it is played.

I've not seen the case of the infinite infect cities loop, but it's possible it's a product of these issues. A quick fix would be to add a second reaction node in the discarding cards step, to be ignored if player doesn't need to discard 2 cards (which is the maximum they'd need to discard anyway).

A proper fix would be rewriting part of the game to use the new ExtendedSequence actions instead, which better handle dynamic actions and complex reactions. I'll have a look at this sometime soon to get it to a better state.

In the mean time, something that might help a bit with debugging: I've pushed a little update to the Pandemic GUI in the `pandemic-update` branch, which has some extra information on it and it should make it more handy to understand what's going on / play the game yourself and test things out.

**Comment by kristian-angelin** [comment]
I've been experimenting with the framework for my undergraduate thesis and I think the main issue is how nodes gets copied when creating a new copy of the forward model. They are shallow copied, allowing changes made during the MCTS-simulation to carry over to the real game state.

I solved the issue by creating copy constructors for the Node class and subclasses.

If there is need for a temporary fix until the ExtendedSequence has been implemented I could set up a pull request.
